### PR TITLE
Support remote debugging for Azure Functions

### PR DESCRIPTION
**Pull request by Flanker32**
What does this implement/fix? Explain your changes.
---------------------------------------------------
…

Does this close any currently open issues?
------------------------------------------

Any relevant logs, screenshots, error output, etc.?
-------------------------------------

Any other comments?
-------------------
…

Has this been tested?
---------------------------
- [ ] Tested

### CODE REVIEW

File: Utils/azure-toolkit-ide-libs/azure-toolkit-ide-appservice-lib/src/main/java/com/microsoft/azure/toolkit/ide/appservice/function/remotedebugging/FunctionPortForwarderWebSocketListener.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import java.nio.channels.WritableByteChannel;
+import java.util.function.BooleanSupplier;
+
+public class FunctionPortForwarderWebSocketListener extends PortForwarderWebSocketListener {
+    public FunctionPortForwarderWebSocketListener(ReadableByteChannel in, WritableByteChannel out, AbstractPortForwarder forwarder) {
+        super(in, out, forwarder);
+    }
+
+    protected void pipe(ReadableByteChannel in, WebSocket webSocket, BooleanSupplier isAlive) throws IOException, InterruptedException {
+        final ByteBuffer buffer = ByteBuffer.allocate(4096);
```
Comment (Flanker32) [inline Utils/azure-toolkit-ide-libs/azure-toolkit-ide-appservice-lib/src/main/java/com/microsoft/azure/toolkit/ide/appservice/function/remotedebugging/FunctionPortForwarderWebSocketListener.java:26]: no need to put buffer header for function/basic web socket listener

File: Utils/azure-toolkit-ide-libs/azure-toolkit-ide-common-lib/src/main/java/com/microsoft/azure/toolkit/ide/common/portforwarder/AbstractPortForwarder.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    public void startForward(final int localPort) {
+        try {
+            this.server = ServerSocketChannel.open().bind(new InetSocketAddress(localPort));
+            final OkHttpClient okHttpClient = new OkHttpClient();
+            final PortForwarderWebSocketListener listener = createWebSocketListener(server.accept());
+            final CompletableFuture<WebSocket> future = createSocketBuilder(okHttpClient).buildAsync(listener);
+            future.whenComplete((socket, throwable) -> Optional.ofNullable(throwable).ifPresent(t -> listener.onError(socket, t)));
+        } catch (final IOException e) {
+            throw new AzureToolkitRuntimeException("Unable to start debugging.", e);
```
Comment (Flanker32) [inline Utils/azure-toolkit-ide-libs/azure-toolkit-ide-common-lib/src/main/java/com/microsoft/azure/toolkit/ide/common/portforwarder/AbstractPortForwarder.java:31]: call stop here once server is started

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/appservice/AppServiceIntelliJActionsContributor.java [hunk 3]
```diff
@@ -185,6 +188,19 @@ public void registerHandlers(AzureActionManager am) {
                     .region(r.getRegion())
                     .resourceGroup(ResourceGroupConfig.fromResource(r)).build());
         am.registerHandler(WebAppActionsContributor.GROUP_CREATE_WEBAPP, (r, e) -> true, groupCreateWebAppHandler);
+
+        final BiPredicate<FunctionAppBase<?, ?, ?>, AnActionEvent> isFunction = (r, e) -> r instanceof FunctionAppBase<?, ?, ?>;
+        final BiConsumer<FunctionAppBase<?, ?, ?>, AnActionEvent> enableRemoteDebuggingHandler = (c, e) ->
+                FunctionEnableRemoteDebuggingAction.enableRemoteDebugging(c, e.getProject());
+        am.registerHandler(FunctionAppActionsContributor.ENABLE_REMOTE_DEBUGGING, isFunction, enableRemoteDebuggingHandler);
+
+        final BiConsumer<FunctionAppBase<?, ?, ?>, AnActionEvent> disableRemoteDebuggingHandler = (c, e) ->
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice/src/main/java/com/microsoft/azure/toolkit/intellij/appservice/AppServiceIntelliJActionsContributor.java:197]: hide enable/disable remote debugging in explorer menu based on status

File: Utils/azure-toolkit-ide-libs/azure-toolkit-ide-appservice-lib/src/main/java/com/microsoft/azure/toolkit/ide/appservice/function/FunctionAppActionsContributor.java [hunk 4]
```diff
@@ -78,6 +80,9 @@ public void registerGroups(AzureActionManager am) {
             ResourceCommonActionsContributor.RESTART,
             ResourceCommonActionsContributor.DELETE,
             "---",
+            FunctionAppActionsContributor.ENABLE_REMOTE_DEBUGGING,
```
Comment (Flanker32) [inline Utils/azure-toolkit-ide-libs/azure-toolkit-ide-appservice-lib/src/main/java/com/microsoft/azure/toolkit/ide/appservice/function/FunctionAppActionsContributor.java:83]: ->
```
            "---",
            FunctionAppActionsContributor.REMOTE_DEBUGGING,
            FunctionAppActionsContributor.ENABLE_REMOTE_DEBUGGING,
            FunctionAppActionsContributor.DISABLE_REMOTE_DEBUGGING,
            "---",
```

@ivywei0125  Please up asa as well

File: Utils/azure-toolkit-ide-libs/azure-toolkit-ide-common-lib/src/main/java/com/microsoft/azure/toolkit/ide/common/portforwarder/AbstractPortForwarder.java [hunk 5]
```diff
@@ -20,9 +20,9 @@
 public abstract class AbstractPortForwarder {
     protected ServerSocketChannel server;
 
-    public void startForward(final int localPort) {
+    public void startForward(ServerSocketChannel server) {
```
Comment (Flanker32) [inline Utils/azure-toolkit-ide-libs/azure-toolkit-ide-common-lib/src/main/java/com/microsoft/azure/toolkit/ide/common/portforwarder/AbstractPortForwarder.java:23]: refactor startForward to 
- start local server
- run `server.accept()` in async thread

**Review by wangmingliang-ms — APPROVED** [review]
let's merge this pr first,  remember to fix comments in later PRs.

**Pull request merged**